src/D2L.CodeStyle.Analyzers/Diagnostics.cs

```diff
@@ -514,5 +514,77 @@ public static class Diagnostics {
 			isEnabledByDefault: true,
 			description: "We require that event types be sealed."
 		);
+
+		public static readonly DiagnosticDescriptor MutableBaseType = new DiagnosticDescriptor(
+			id: "D2L0061",
+			title: "Base class for immutable type must be [Immutable] (or [ImmutableBaseClass])",
+			messageFormat: "{0}'s base class is {1} which is missing [Immutable] (or, more weakly, [ImmutableBaseClass]).",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor ArraysAreMutable = new DiagnosticDescriptor(
+			id: "D2L0062",
+			title: "Arrays are mutable and thus can't be held by an immutable type.",
+			messageFormat: "{0}[]'s are mutable even if {0}'s are immutable because array items can be substituted. Prefer an immutable container type such as ImmutableArray<{0}>.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor DynamicObjectsAreMutable = new DiagnosticDescriptor(
+			id: "D2L0063",
+			title: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			messageFormat: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeParameterIsNotKnownToBeImmutable = new DiagnosticDescriptor(
+			id: "D2L0064",
+			title: "Type parameter is not known to be immutable.",
+			messageFormat: "The type parameter {0} is not known to be immutable. Add [Immutable] to it.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor UnexpectedTypeKind = new DiagnosticDescriptor(
+			id: "D2L0065",
+			title: "Unsupported type kind inside an immutable type declaration.",
+			messageFormat: "The type kind {0} is unexpected/unhandled by the immutability analyzer, and for safety we are emitting an error. This is a bug in the analyzer.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeNeedsImmutableAttribute = new DiagnosticDescriptor(
+			id: "D2L0066",
+			title: "Type may be mutable.",
+			messageFormat: "The {0} {1} is missing the [Immutable]{2} attribute, so it isn't safe to be held by an immutable type",
```

Comment:
"The interface IFoo is missing the [Immutable] attribute, so it isn't safe to be held by an immutable type."
"The class Bar is missing the [Immutable] (or [ImmutableBaseClass]) attribute, so it isn't safe to be held by an immutable type."

Isn't there already an error for base classes? Yes. But the logic for `[ImmutableBaseClass]` is used for other stuff too (it's a very bad name). e.g.:

```
private static readonly IFoo m_foo = new Foo( ... );
```

In this case (modulo the #190 bug) things are allowed if `Foo` has `[ImmutableBaseClass]`.

(I'd like to change the attribute up a bit, but I'll leave that talk for the next PR (which doesn't touch the attribute but will better contextualize the discussion))